src/gluonnlp/model/bert.py

```diff
@@ -405,22 +410,33 @@ def _apply_pooling(self, sequence):
         outputs = sequence[:, 0, :]
         return self.pooler(outputs)
 
-    def _decode(self, sequence, valid_length=None):
+    def _decode(self, sequence, positions):
         """Generate unormalized prediction for the masked language model task.
-
         This is only used for pre-training the BERT model.
+
+        Inputs:
+            - **sequence**: input tensor of sequence encodings.
+              Shape (batch_size, seq_length, units).
+            - **position**: input tensor of position of tokens for decoding.
+              Shape (batch_size, max_num_positions)
+
+        Outputs:
+            - **decode_outputs**: output tensor of token predictions for target positions.
+                Shape (batch_size, max_num_positions, vocab_size).
         """
-        if valid_length is None:
-            last_step = sequence[:, -1, :]
-        else:
-            batch_size = sequence.shape[0]
-            ctx = valid_length.context
-            dtype = valid_length.dtype
-            batch_idx = mx.nd.arange(0, batch_size, dtype=dtype, ctx=ctx).reshape((1, -1))
-            valid_length = (valid_length - 1).reshape((1, -1))
-            last_step_idx = mx.nd.Concat(batch_idx, valid_length, dim=0)
-            last_step = mx.nd.gather_nd(sequence, last_step_idx)
-        return self.decoder(last_step)
+        batch_size = sequence.shape[0]
+        max_num_positions = positions.shape[1]
+        ctx = positions.context
+        dtype = positions.dtype
+        # batch_idx = [0,0,0,1,1,1,2,2,2...]
+        # positions = [1,2,4,0,3,4,2,3,5...]
+        batch_idx = mx.nd.arange(0, batch_size, repeat=max_num_positions, dtype=dtype, ctx=ctx)
+        batch_idx = batch_idx.reshape((1, -1))
+        positions = positions.reshape((1, -1))
+        position_idx = mx.nd.Concat(batch_idx, positions, dim=0)
+        encoded = mx.nd.gather_nd(sequence, position_idx)
```

Comment:
This bunch of codes can be expressed with simple one line... 😃 

`encoded = sequence.take(positions, axis=1).squeeze(axis=1)`

Reply:
I think, `nn.Dense`  needs to be with `flatten=False` to keep timesteps. 

```
     def _get_decoder(self, units, vocab_size, embed, prefix):
         """ Construct a decoder for the masked language model task """
         with self.name_scope():
             decoder = nn.HybridSequential(prefix=prefix)
             decoder.add(nn.Dense(units, flatten=False))
             decoder.add(GELU())
             decoder.add(BERTLayerNorm(in_channels=units))
             decoder.add(nn.Dense(vocab_size, flatten=False, params=embed.params))
         return decoder
return decoder
```

Reply:
Thanks for the suggestions! Adding the flatten flag makes sense. I am not able to follow the first comment though. I tried the following code:
```
import mxnet as mx
seq = mx.nd.arange(0, 60).reshape((3,4,5))
print(seq)
positions = mx.nd.array([0,1,0,3,1,2,1,2]).reshape((3,2))
print(positions)
encoded = mx.nd.take(seq, positions, axis=1)
print(encoded)
```
which prints
```
[[[ 0.  1.  2.  3.  4.]
  [ 5.  6.  7.  8.  9.]
  [10. 11. 12. 13. 14.]
  [15. 16. 17. 18. 19.]]

 [[20. 21. 22. 23. 24.]
  [25. 26. 27. 28. 29.]
  [30. 31. 32. 33. 34.]
  [35. 36. 37. 38. 39.]]

 [[40. 41. 42. 43. 44.]
  [45. 46. 47. 48. 49.]
  [50. 51. 52. 53. 54.]
  [55. 56. 57. 58. 59.]]]
<NDArray 3x4x5 @cpu(0)>

[[0. 1.]
 [0. 3.]
 [1. 2.]]
<NDArray 3x2 @cpu(0)>

[[[[ 0.  1.  2.  3.  4.]
   [ 5.  6.  7.  8.  9.]]

  [[ 0.  1.  2.  3.  4.]
   [15. 16. 17. 18. 19.]]

  [[ 5.  6.  7.  8.  9.]
   [10. 11. 12. 13. 14.]]]


 [[[20. 21. 22. 23. 24.]
   [25. 26. 27. 28. 29.]]

  [[20. 21. 22. 23. 24.]
   [35. 36. 37. 38. 39.]]

  [[25. 26. 27. 28. 29.]
   [30. 31. 32. 33. 34.]]]


 [[[40. 41. 42. 43. 44.]
   [45. 46. 47. 48. 49.]]

  [[40. 41. 42. 43. 44.]
   [55. 56. 57. 58. 59.]]

  [[45. 46. 47. 48. 49.]
   [50. 51. 52. 53. 54.]]]]
<NDArray 3x3x2x5 @cpu(0)>
```
This gives me a 4-D array of shape (3,3,2,5). But what I intended to do is to select different position for each sample in the batch. So I am expecting a result of (3,2,5)

```
[[[ 0  1  2  3  4]
  [ 5  6  7  8  9]]

 [[20 21 22 23 24]
  [35 36 37 38 39]]

 [[45 46 47 48 49]
  [50 51 52 53 54]]]
<NDArray 3x2x5 @cpu(0)>
```
Did I miss something?

Reply:
Oh, you are right, I was confused and didn't test fully. Thanks.

Reply:
That's ok ^_^
The output of gather_nd is 2-D, that's why setting `flatten=False/True` in the last dense layer doesn't make any difference.

Reply:
Right. Shape will be ok, but 3D reshaping needs before `self.encoder()` because maskedLM sequences have no dependency between [MASK]. 
I think, it still need to keep `flatten=False`.

How do you think?

```
        batch_size, _, hidden_size = sequence.shape
        max_num_positions = positions.shape[1]
        ctx = positions.context
        dtype = positions.dtype
        # batch_idx = [0,0,0,1,1,1,2,2,2...]
        # positions = [1,2,4,0,3,4,2,3,5...]
        batch_idx = mx.nd.arange(0, batch_size, repeat=max_num_positions, dtype=dtype, ctx=ctx)
        batch_idx = batch_idx.reshape((1, -1))
        positions = positions.reshape((1, -1))
        position_idx = mx.nd.Concat(batch_idx, positions, dim=0)
        encoded = mx.nd.gather_nd(sequence, position_idx).reshape((batch_size, max_num_positions, hidden_size))
        decoded = self.decoder(encoded)
        return decoded
```

Reply:
That makes sense! I'll push an update